### ISSUE TITLE
Components as dictionary keys do not round-trip through JSON

### ISSUE THREAD

**Issue by BeksOmega** [mit-cml/appinventor-sources#2221]
**Describe the bug**
If you add a component as a dictionary key, then serialize and deserialize the dictionary, you can no longer use the component to access values in the dictionary.

**Affects**

- [ ] Designer
- [X] Blocks editor
- [X] Companion
- [ ] Compiled apps
- [ ] Buildserver
- [ ] Debugging
- [ ] Other... (please describe)

**Expected behavior**
I would still be able to use components to access values in the dictionary.

**Steps to reproduce**

Using the following blocks the first label correctly receives "label 1 text" while the second receives "not found".
![JsonBlocks](https://user-images.githubusercontent.com/25440652/86054585-44d05080-ba0f-11ea-92d9-812f55198080.png)
You can also load this zipped .aia: [test2.zip](https://github.com/mit-cml/appinventor-sources/files/4847953/test2.zip)

**Comment by ewpatton** [comment]
App Inventor has no theory about serializing components, so this is to be expected. One challenge if we do eventually allow for serializing components is it might have security implications when used with components like CloudDB.

**Comment by halatmit** [comment]
I agree that serializing components could be tricky and that we should stay
away from this for now.

== Hal

Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.

**Comment by BeksOmega** [comment]
I wonder if we could fake it so that from the perspective of people interacting with the blocks editor, it looks like the component is serialized, when really we're just serializing something that says "hey dictionary, look here to find that component".

For example if components had some sort of stringify-able UID, we could store that as a value in the Yail-side dictionary instead of storing the actual component. Whenever people try to access that value we would do a lookup and return the component associated with the UID (if the component exists).

The same thing would work for keys, just store the UID as the key rather than the actual component.

All that being said, if that's even possible it's probably too much work hehe. I just wanted to throw the idea out there.

**Comment by ewpatton** [comment]
Some form of UID would be good potentially. We have one on the design side but it's not included in the code gen. Some possible edge cases we'd have to think about include:

In the companion, I load project A and then serialize some data with Label1 to TinyDB
I open project B and read out data from TinyDB, but I don't have Label1. Or, by random chance that project has a different component with the same UID.

If we're not using UIDs, do we care to try and reconcile when components are renamed?

Currently, every screen has UID 0, so we'd also have to address that.

Maybe a URI-like scheme? appinventor-component://Screen1/Label1 so there's less chance of collisions?

There's likely more but that's an initial brain dump.

**Comment by ewpatton** [comment]
Another potential problem is that people might expect if they serialize "the component" that they are serializing its state rather than its identifier. We see this with people trying to write, say, a TextBox or Label to a TinyDB and then having the string come back as the Java serialized name of the component instance rather than the text of the textbox/label. Implementing this might cause further confusion.

**Comment by BeksOmega** [comment]
> In the companion, I load project A and then serialize some data with Label1 to TinyDB
I open project B and read out data from TinyDB, but I don't have Label1. Or, by random chance that project has a different component with the same UID.

I think different people will probably have different preferences for this. But I like the behavior where it would just return the "not found" option. Or in the case that the dictionary is converted to a list, the Label1 value would be turned into some kind of falsy value (null, the empty list, etc).

> Maybe a URI-like scheme? appinventor-component://Screen1/Label1 so there's less chance of collisions?

I really like this!

> If we're not using UIDs, do we care to try and reconcile when components are renamed?

You mean like if they serialize some data, change the names, and then try to deserialize the data? ...This is tricky.

So on the one hand if we use names and not UIDs then if someone deletes a component, and brings it back by recreating it with the same name (but it's a different instance), they can still use that component (which looks identical to them) to access the old data. Essentially we handle them changing the UID but not them changing the name.

On the other hand if we use UIDs and not names then if someone does the procedure described above (serialize data, change names, then deserialize data) they can still use the renamed component to access the old data. Essential we handle them changing the name but not the UID.

Dar n this may be an unwinnable situation... Ideally we would want to handle changing either piece of data (because that's what makes sense to the user) but I don't know how we could build flexibility on both ends.

> Another potential problem is that people might expect if they serialize "the component" that they are serializing its state rather than its identifier. We see this with people trying to write, say, a TextBox or Label to a TinyDB and then having the string come back as the Java serialized name of the component instance rather than the text of the textbox/label. Implementing this might cause further confusion.

I get where you're coming from, but I think this problem already exists. Say we have two users:

User 1 serializes a TextBox and expects to get the state/string back. But they actually get a string like "com.google.appinventor.components.runtime.TextBox@18deebd"

User 2 serializes a TextBox and expects to get the TextBox back. But they  actually get a string like "com.google.appinventor.components.runtime.TextBox@18deebd"

We can't fix user 1's situation, but we can fix user 2. That's just my way of looking at it though.

And again, adding this may not be worth it hehe. Serializing components as keys or values is arguably a pretty slim use-case.

**Comment by BeksOmega** [comment]
Was talking to Mr. Lyn about this yesterday and he also suggested just throwing an error/warning if a user tries to serialize a dictionary that contains things which cannot be serialized. This would at least inform the users that think components can round-tripped that they can't. And if we just threw an error we wouldn't have to worry about the edge cases of deserializing a component reference into a place where the component no longer exists.

**Comment by ewpatton** [comment]
That’s probably better for now. We can always relax the restriction in the future. However, we should still have a way of serializing them for display (e.g., Do It) even if they won’t round-trip.

**Comment by halatmit** [comment]
sound like a good way to go, at least at the start

== Hal

Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
